docs/guides/dropout.rst

```diff
@@ -0,0 +1,299 @@
+  from flax.training import train_state
+
+  class TrainState(train_state.TrainState): #!
+    key: jax.random.KeyArray #!
+
+  state = TrainState.create( #!
+    apply_fn=my_model.apply,
+    params=params,
+    key=dropout_key, #!
+    tx=optax.adam(1e-3)
+  )
+
+* Next, in the Flax training step function, ``train_step``, generate a new PRNG
+  key from the ``dropout_key`` to apply dropout at each step. This can be done with one of the following:
+
+  * `jax.random.split() <https://jax.readthedocs.io/en/latest/_autosummary/jax.random.split.html>`__; or
+  * `jax.random.fold_in() <https://jax.readthedocs.io/en/latest/_autosummary/jax.random.fold_in.html>`__
+
+  Using ``jax.random.fold_in()`` is generally faster. When you use
+  ``jax.random.split()`` you split off a PRNG key that can be reused
+  afterwards. However, using ``jax.random.fold_in()`` makes sure to 1) fold in
+  unique data; and 2) can result in longer sequences of PRNG streams.
+
+* Finally, when performing the forward pass, pass the new PRNG key to ``state.apply_fn()``
+  as an extra parameter.
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  @jax.jit
+  def train_step(state: TrainState, batch):
+
+    def loss_fn(params):
+      logits = state.apply_fn(
+        {'params': params},
+        x=batch['image'],
+
+
+        )
+      loss = optax.softmax_cross_entropy_with_integer_labels(
+        logits=logits, labels=batch['label'])
+      return loss, logits
+    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+    (loss, logits), grads = grad_fn(state.params)
+    state = state.apply_gradients(grads=grads)
+    return state
+
+  ---
+  @jax.jit
+  def train_step(state: TrainState, batch, dropout_key): #!
+    dropout_train_key = jax.random.fold_in(key=dropout_key, data=state.step) #!
+    def loss_fn(params):
+      logits = state.apply_fn(
+        {'params': params},
+        x=batch['image'],
+        training=True, #!
+        rngs={'dropout': dropout_train_key} #!
+        )
+      loss = optax.softmax_cross_entropy_with_integer_labels(
+        logits=logits, labels=batch['label'])
+      return loss, logits
+    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+    (loss, logits), grads = grad_fn(state.params)
+    state = state.apply_gradients(grads=grads)
+    return state
+
+Flax examples with dropout
+**************************
+
+* A `Transformer-based model <https://github.com/google/flax/blob/main/examples/wmt/models.py>`__
+  trained on the WMT Machine Translation dataset. This example uses dropout and attention dropout.
+
+More Flax examples that use randomness
+**************************************
+
+* Applying word dropout to a batch of input IDs in a
+  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
+  context.
+
+* Defining a prediction token in a decoder of a
+  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.
```

Comment:
Small suggestion: "using randomness" is a bit odd since any ML model uses randomness when initializing the parameters. Also the sst model actually uses dropout, but a custom one, so i would put it under the dropout section.

```suggestion
Flax examples with dropout
**************************

* A `Transformer-based model <https://github.com/google/flax/blob/main/examples/wmt/models.py>`__
  trained on the WMT Machine Translation dataset. This example uses dropout and attention dropout.
  
* Applying word dropout to a batch of input IDs in a
  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
  context. This example defines a custom Dropout layer.

More Flax examples that use `make_rng()`
**************************************

* Defining a prediction token in a decoder of a
  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.
```

Reply:
Done:

```
Flax examples with dropout
**************************

* A `Transformer-based model <https://github.com/google/flax/blob/main/examples/wmt/models.py>`__
  trained on the WMT Machine Translation dataset. This example uses dropout and attention dropout.

* Applying word dropout to a batch of input IDs in a
  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
  context. This example uses a custom :meth:`flax.linen.Dropout` layer.

More Flax examples that use Module ``make_rng()``
*************************************************

* Defining a prediction token in a decoder of a
  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.

```

Thanks @marcvanzee 

Also changed the previous section title from "Training" to "`TrainState` and the training step".